engine/consensus/matching/engine.go

```diff
@@ -98,1063 +75,169 @@ func New(
-	added, err := e.approvals.Add(approval)
-	if err != nil {
-		return err
-	}
-	if !added {
-		log.Debug().Msg("skipping approval already in mempool")
-		return nil
-	}
-	e.mempool.MempoolEntries(metrics.ResourceApproval, e.approvals.Size())
-
-	return nil
-}
-
-// checkSealing checks if there is anything worth sealing at the moment.
-func (e *Engine) checkSealing() {
-	// only check sealing when no one else is checking
-	canCheck := e.isCheckingSealing.CAS(false, true)
-	if !canCheck {
-		return
-	}
-
-	defer e.isCheckingSealing.Store(false)
-
-	e.unit.Lock()
-	defer e.unit.Unlock()
-
-	err := e.checkingSealing()
-	if err != nil {
-		e.log.Fatal().Err(err).Msg("error in sealing protocol")
-	}
-}
-
-func (e *Engine) checkingSealing() error {
-	startTime := time.Now()
-	sealingSpan, _ := e.tracer.StartSpanFromContext(context.Background(), trace.CONMatchCheckSealing)
-	defer func() {
-		e.metrics.CheckSealingDuration(time.Since(startTime))
-		sealingSpan.Finish()
-	}()
-
-	sealableResultsSpan := e.tracer.StartSpanFromParent(sealingSpan, trace.CONMatchCheckSealingSealableResults)
-
-	// get all results that have collected enough approvals on a per-chunk basis
-	sealableResults, err := e.sealableResults()
-	if err != nil {
-		return fmt.Errorf("could not get sealable execution results: %w", err)
-	}
-	if len(sealableResults) == 0 {
-		// skip if no results can be sealed yet
-		return nil
-	}
-	e.log.Info().Int("num_results", len(sealableResults)).Msg("identified sealable execution results")
-
-	// log warning if we are going to overflow the seals mempool
-	if space := e.seals.Limit() - e.seals.Size(); len(sealableResults) > int(space) {
-		e.log.Warn().
-			Int("space", int(space)).
-			Msg("overflowing seals mempool")
-	}
-
-	// Start spans for tracing within the parent spans trace.CONProcessBlock and
-	// trace.CONProcessCollection
-	for _, incorporatedResult := range sealableResults {
-		// For each execution result, we load the trace.CONProcessBlock span for the executed block. If we find it, we
-		// start a child span that will run until this function returns.
-		if span, ok := e.tracer.GetSpan(incorporatedResult.Result.BlockID, trace.CONProcessBlock); ok {
-			childSpan := e.tracer.StartSpanFromParent(span, trace.CONMatchCheckSealing, opentracing.StartTime(startTime))
-			defer childSpan.Finish()
-		}
-
-		// For each execution result, we load all the collection that are in the executed block.
-		index, err := e.indexDB.ByBlockID(incorporatedResult.Result.BlockID)
-		if err != nil {
+	for {
+		pendingReceipt, receiptSink := fetchEvent(&c.pendingReceipts, c.receiptSink)
+		pendingApproval, approvalSink := fetchEvent(&c.pendingApprovals, c.approvalSink)
+		pendingApprovalResponse, approvalResponseSink := fetchEvent(&c.pendingApprovalResponses, c.approvalResponseSink)
+		select {
+		case event := <-c.pendingEventSink:
+			c.processPendingEvent(event)
+		case receiptSink <- pendingReceipt:
```

Comment:
this will block unless receiptSink is a buffered channel

Reply:
That's the idea that it won't, with this select it will block on whole statement and will unblock in 3 cases:
1. We are quitting
2. We are able to send data
3. We are able to receive data. 

So even if we can't send data to core we are still able to get data from even sink.

Reply:
I am not sure about that. 
I think it wont process the next event unil receiptSink <- pendingReceipt unblocks, whcih happens only if core consumes the receipt immediately. 
But I might be wrong..would be interesting to confirm with a test

Reply:
I have updated PR description, you can find there a link with working example why this works.

Reply:
Yeah thanks a lot for the example. I wasn't aware that one could do non-blocking sends like that!
Very nice!